// uport-connect/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MNID = exports.Credentials = exports.SimpleSigner = exports.QRUtil = exports.ConnectCore = exports.Connect = undefined;

var _Connect = require('./Connect');

var _Connect2 = _interopRequireDefault(_Connect);

var _ConnectCore = require('./ConnectCore');

var _ConnectCore2 = _interopRequireDefault(_ConnectCore);

var _qrdisplay = require('./util/qrdisplay');

var _mnid = require('mnid');

var _uport = require('uport');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var QRUtil = { getQRDataURI: _qrdisplay.getQRDataURI, closeQr: _qrdisplay.closeQr, openQr: _qrdisplay.openQr };

var MNID = { encode: _mnid.encode, decode: _mnid.decode, isMNID: _mnid.isMNID };
exports.Connect = _Connect2.default;
exports.ConnectCore = _ConnectCore2.default;
exports.QRUtil = QRUtil;
exports.SimpleSigner = _uport.SimpleSigner;
exports.Credentials = _uport.Credentials;
exports.MNID = MNID;
